packages/core/echo/echo-db/src/proxy-db/mutable-schema-registry.ts

```diff
@@ -19,17 +21,167 @@ import { createStoredSchema } from '@dxos/live-object';
+   * schema with the same name and version already exists, the method does nothing.
+   * If a different schema with the same name and version exists, the method throws
+   * an error. If no schema with the same name and version exists, a new schema will
+   * be inserted based on semantic versioning rules.
+   */
+  register(input: RegisterSchemaInput[]): Promise<SchemaRecord[]>;
+
+  /**
+   * @deprecated Use `query()`.
+   */
   getSchema(typename: string): MutableSchema | undefined;
-  query(): Promise<MutableSchema[]>;
+
+  /**
+   * @deprecated Use `query()`.
+   */
   subscribe(cb: SchemaSubscriptionCallback): UnsubscribeCallback;
 }
 
+export type SchemaRegistryQuery = {
+  /**
+   * Filter by schema ID.
+   * Schema id is a DXN with `echo` or `type` kind.
+   */
+  id?: string[];
+
+  /**
+   * One or more typenames to filter by.
+   */
+  typename?: string[];
+
+  /**
+   * [Semver Range](https://docs.npmjs.com/cli/v6/using-npm/semver#ranges) for the schema version.
+   */
+  version?: string;
+};
+
+export interface SchemaRegistryLiveQuery<T> {
+  /**
+   * Returns query results synchronously.
+   * Supports signals notifications.
+   * User must call `subscribe` for reactive notifications to be enabled.
+   */
+  get results(): T[];
+
+  /**
+   * Runs the query and returns all results.
+   */
+  run(): Promise<T[]>;
+
+  /**
+   * Runs the query and returns first result.
+   *
+   * @throws If query returns 0 entries.
+   */
+  first(): Promise<T>;
+
+  /**
+   * Subscribe to the query results reactively.
+   * Enables signals notifications for `results`.
+   */
+  subscribe(cb: (self: this) => void, opts: { fire?: boolean }): UnsubscribeCallback;
+}
+
+/**
+ * Input for schema registration.
+ * Either one of the schema variants must be provided.
+ * The typename, version and schema mutability metadata is read from the schema annotations.
+ */
+export type RegisterSchemaInput = {
+  /**
+   * Schema to register in the Effect format.
+   */
+  schema?: AnyEchoObjectSchema;
+
+  /**
+   * Schema to register in the JSON Schema format.
+   */
+  jsonSchema?: JsonSchemaType;
+};
+
+export type AnyEchoObjectSchema = S.Struct<{ [key: string]: S.Schema.AnyNoContext }>;
+
+/**
+ * Record of the schema stored in the registry
+ */
+// TODO(dmaretskyi): Replaces MutableSchema, StaticSchema.
+export interface SchemaRecord {
+  /**
+   * String identifier for the schema.
+   * In practice it's an `echo` or `type` DXN.
+   */
+  get id(): string;
+
+  get mutable(): boolean;
+
+  /**
+   * @returns Rendered effect schema snapshot for this entry.
+   */
+  getSchema(): AnyEchoObjectSchema;
+
+  /**
+   * @returns This schema snapshot in JSON-schema format.
+   */
+  getJsonSchema(): JsonSchemaType;
+
+  /**
+   * Get backing ECHO object for this schema.
+   */
+  getBackingObject(): Promise<ReactiveEchoObject<StoredSchema> | undefined>;
+
+  /**
+   * Updates typename.
+   *
+   * Only callable if the schema is mutable.
```

Comment:
`@throws Error if schema is not mutable` (jsdoc)

Reply:
this is fine, but it might be more precise to have update verbs on the registry?

registry.update(schema, {
  upsert: Fields,
  rename: [{ from, to }],
  delete: string[],
});

OR perhaps just set the new struct/schema directly (not fine grained?)